influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -108,59 +115,208 @@ struct CacheValue {
+        let map = self.map.read().await;
+        let entry = map.get(path)?;
+        match &entry.state {
+            CacheEntryState::Fetching(tx) => {
+                let mut rx = tx.subscribe();
+                // TODO(trevor): is it possible that the sender has sent the result at this point?
+                // if so we need to check the rx first and maybe use more sophisticated state than
+                // an Option in the watch channel type.
+                rx.changed().await.ok()?;
+                let v = rx.borrow_and_update().clone();
+                v
+            }
+            CacheEntryState::Success(v) => {
+                entry.increment_hits();
+                Some(Arc::clone(v))
+            }
+        }
     }
 
-    fn keep(&self) -> bool {
-        self.is_fetching() || self.is_success()
+    /// Check if an entry in the cache is in process of being fetched or if it was already fetched
+    /// successfully
+    async fn path_already_fetched(&self, path: &Path) -> bool {
+        self.map.read().await.get(path).is_some()
     }
-}
 
-/// Implements the [`WeightScale`] trait to determine a [`CacheEntry`]'s size on insertion to
-/// the cache
-#[derive(Debug)]
-struct CacheEntryScale;
+    /// Insert a `Fetching` entry to the cache with a watcher
+    async fn set_fetching(&self, path: &Path) {
+        let (tx, _) = watch::channel(None);
+        let entry = CacheEntry {
+            state: CacheEntryState::Fetching(tx),
+            hits: AtomicI8::new(-1),
+        };
+        let additional = entry.size();
+        self.map.write().await.insert(path.clone(), entry);
+        self.used.fetch_add(additional, Ordering::SeqCst);
+    }
+
+    /// Update a `Fetching` entry to a `Success` entry in the cache
+    async fn set_success(&self, path: &Path, value: CacheValue) -> Result<(), anyhow::Error> {
+        match self.map.write().await.entry(path.clone()) {
+            Entry::Occupied(mut o) => {
+                let entry = o.get_mut();
+                let tx = match &entry.state {
+                    CacheEntryState::Fetching(tx) => tx.clone(),
+                    _ => bail!("attempted to set success state on a non-fetching cache entry"),
+                };
+                let cache_value = Arc::new(value);
+                entry.state = CacheEntryState::Success(Arc::clone(&cache_value));
+                entry.hits.store(1, Ordering::SeqCst);
+                // TODO(trevor): what if size is greater than cache capacity?
+                let additional = entry.size();
+                self.used.fetch_add(additional, Ordering::SeqCst);
+                let _ = tx.send(Some(cache_value));
+                Ok(())
+            }
+            Entry::Vacant(_) => bail!("attempted to set success state on an empty cache entry"),
+        }
+    }
+
+    /// Remove an entry from the cache, as well as its associated size from the used capacity
+    async fn remove(&self, path: &Path) {
```

Comment:
May be worth noting that this is _O(N)_ due to the underlying `IndexMap`, but generally, remove operations would occur seldom, since it is only on object store delete.